File: src/main/java/com/fengsheng/handler/use_li_you_tos.java
package com.fengsheng.handler;

import com.fengsheng.HumanPlayer;
import com.fengsheng.Player;
import com.fengsheng.card.Card;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class use_li_you_tos extends AbstractProtoHandler<Fengsheng.use_li_you_tos> {
    private static final Logger log = Logger.getLogger(use_li_you_tos.class);

    @Override
    protected void handle0(HumanPlayer r, Fengsheng.use_li_you_tos pb) {
        if (!r.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + r.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        Card card = r.findCard(pb.getCardId());
        if (card == null) {
            log.error("没有这张牌");
            return;
        }
        if (card.getType() != Common.card_type.Li_You) {
            log.error("这张牌不是利诱，而是" + card);
            return;
        }
        if (pb.getPlayerId() < 0 || pb.getPlayerId() >= r.getGame().getPlayers().length) {
            log.error("目标错误: " + pb.getPlayerId());
            return;
        }
        Player target = r.getGame().getPlayers()[r.getAbstractLocation(pb.getPlayerId())];
        if (card.canUse(r.getGame(), r, target)) {
            r.incrSeq();
            card.execute(r.getGame(), r, target);
        }
    }
}

File: src/main/kotlin/handler/use_li_you_tos.kt
package com.fengsheng.handler

import com.fengsheng.HumanPlayer
import com.fengsheng.protos.Common.card_type.Li_You
import com.fengsheng.protos.Fengsheng
import org.apache.logging.log4j.kotlin.logger

class use_li_you_tos : AbstractProtoHandler<Fengsheng.use_li_you_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.use_li_you_tos) {
        if (!r.checkSeq(pb.seq)) {
            logger.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        var card = r.findCard(pb.cardId)
        if (card == null) {
            logger.error("没有这张牌")
            r.sendErrorMessage("没有这张牌")
            return
        }
        if (card.type != Li_You) {
            logger.error("这张牌不是利诱，而是$card")
            r.sendErrorMessage("这张牌不是利诱，而是$card")
            return
        }
        if (pb.playerId < 0 || pb.playerId >= r.game!!.players.size) {
            logger.error("目标错误: ${pb.playerId}")
            r.sendErrorMessage("目标错误: ${pb.playerId}")
            return
        }
        val target = r.game!!.players[r.getAbstractLocation(pb.playerId)]!!
        if (card.type != Li_You) card = card.asCard(Li_You)
        if (card.canUse(r.game!!, r, target)) {
            r.incrSeq()
            card.execute(r.game!!, r, target)
        }
    }
}